### PR TITLE
feat(gossip): schema cache anti-entropy service

### PR DESCRIPTION
**Pull request by domodwyer**
This PR defines the proto types/methods necessary to implement eventual consistency / anti-entropy of the router schema caches between gossip peers.

---

* feat(gossip): schema cache anti-entropy service (4ec07b5d8)
      
      Defines an RPC service to be used by two peers to converge their schema
      cache content by exchanging their serialised Merkle Search Tree pages (a
      compact representation of the MST itself).
      
      This will be used in the latter half of the following sync protocol:
      
                          ┌─────┐                   ┌────┐
                          │Local│                   │Peer│
                          └──┬──┘                   └─┬──┘
                             │ [1] Send content hash  │
                             │────────────────────────>
                             │                        │
                             │                 ┌───────────────┐
                             │                 │Compute hash,  │
                             │                 │stop if equal  │
                             │                 └───────────────┘
                             │                        │
                             │   ╔════════════════╗   │
      ═══════════════════════╪═══╣ Switch to gRPC ╠═══╪═══════════════════════
                             │   ╚════════════════╝   │
                             │                        │
                             │[2] Serialised MST pages│
                             │<────────────────────────
                             │                        │
                      ┌──────────────┐                │
                      │Perform diff  │                │
                      └──────────────┘                │
                             │ [3] Inconsistent pages │
                             │────────────────────────>
                             │                        │
                             │                        │
                ╔═══════╤════╪════════════════════════╪════════════╗
                ║ LOOP  │  For each inconsistent page │            ║
                ╟───────┘    │                        │            ║
                ║            │      [4] Scheams       │            ║
                ║            │<────────────────────────            ║
                ╚════════════╪════════════════════════╪════════════╝
                          ┌──┴──┐                   ┌─┴──┐
                          │Local│                   │Peer│
                          └─────┘                   └────┘
      
      The initial consistency probe request [1] is sent over gossip and is
      used to trigger a further sync of inconsistent MST content if necessary.
      This message is a no-op if the MSTs are found to be fully consistent.
      
      If an inconsistency is detected between the two peers, the protocol
      switches to perform RPC over TCP, calling the AntiEntropyService defined
      in this commit to complete the sync process.
      
      The receiver of the consistency probe [1] calls GetTreeDiff and provides
      the their MST pages [2], causing the local node to compute the diff
      between the two MSTs, and return the set of inconsistent ranges [3] that
      require convergence.
      
      Once the set of inconsistent ranges have been identified, the peer pulls
      all the schemas within those ranges and merges them into the local cache
      to ensure it has all the content of the source node.
      
      Once this protocol has run in both directions between two peers (and in
      absence of further updates between runs) then these two peers are
      guaranteed to have converged.

### CODE REVIEW

File: generated_types/protos/influxdata/iox/gossip/v1/schema_sync.proto [hunk 1]
```diff
@@ -2,6 +2,89 @@ syntax = "proto3";
...
+  string max = 2;
+
+  // A 16-byte MST page hash covering all entries in this page.
+  bytes page_hash = 3;
+}
+
+// An inclusive range of namespace names which contains at least one
+// inconsistent schema.
+message DiffRange {
+  // Minimum inconsistent namespace name in this diff range (inclusive).
```
Comment (savage-engineer) [inline generated_types/protos/influxdata/iox/gossip/v1/schema_sync.proto:52]: Do the minimum and maximum here refer to the lexicographical ordering of namespace names?

**Comment by savage-engineer** [comment]
> ```
>                    ┌─────┐                   ┌────┐
>                    │Local│                   │Peer│
>                    └──┬──┘                   └─┬──┘
>                       │ [1] Send content hash  │
>                       │────────────────────────>
>```

Will the next occurrence of step (1) be blocked on the previous iteration of the convergence protocol completing?

File: generated_types/protos/influxdata/iox/gossip/v1/schema_sync.proto (hunk 1 above)
Comment (domodwyer) [reply generated_types/protos/influxdata/iox/gossip/v1/schema_sync.proto:52]: They do - I legit nearly wrote it, and then thought no one would ever question it 😂 

Happy to put it up in there :+1:

**Comment by domodwyer** [comment]
> Will the next occurrence of step (1) be blocked on the previous iteration of the convergence protocol completing?

I intend to bound the number of in-flight sync rounds, but also prevent more than one sync round being started against any one peer (potentially redundant or overlapping syncs).

**Comment by savage-engineer** [comment]
> I intend to bound the number of in-flight sync rounds, but also prevent more than one sync round being started against any one peer (potentially redundant or overlapping syncs).

Grand, thanks. I was concerned about that multiple syncs against a single peer being gnarly!

**Review by savage-engineer — APPROVED** [review]

**Comment by domodwyer** [comment]
> I was concerned about that multiple syncs against a single peer being gnarly!

2 concurrent syncs to the same peer would potentially re-sync the same keys, which would be inefficient, but it wouldn't affect correctness as they're all monotonic types - keeping the cache entries monotonic makes this whole eventual consistency thing much easier to reason about 🎉

**Pull request merged**